# 异常

## 异常的体系结构

异常的层次结构如下所示：

![img](https://images2015.cnblogs.com/blog/690102/201607/690102-20160728164909622-1770558953.png)

`Throwable`分成两个不同的分支，一个是`Error`，表示**不希望被程序捕获**，或者是程序**无法处理的错误**。另一个是`Exception`，表示程序可能捕获的异常情况或者是程序**可以处理的**异常。

* `Error`: Error类对象都是由Java虚拟机生成并被抛出，大多数错误与代码编写者所执行操作无关。
* Exception：在`Exception`分支中有一个重要的子类`RuntimeException`（运行时异常），该类型的异常自动为你所编写的程序定义`ArrayIndexOutOfBoundsException`（数组下标越界）、`NullPointerException`（空指针异常）、`ArithmeticException`（算术异常）、`MissingResourceException`（丢失资源）、`ClassNotFoundException`（找不到类）等异常，**这些异常是不检查异常，程序中可以选择捕获处理，也可以不处理。这些异常一般是由程序逻辑错误引起的，程序应该从逻辑角度尽可能避免这类异常的发生**；而`RuntimeException`之外的异常我们统称为非运行时异常，类型上属于`Exception`类及其子类，从程序语法角度讲是必须进行处理的异常，如果不处理，程序就不能编译通过。如`IOException`、`SQLException`等以及用户自定义的`Exception`异常，一般情况下不自定义检查异常。



## 检查异常：

在正确的程序运行过程中，很容易出现的、情理可容的异常状况，在一定程度上这种异常的发生是可以预测的，并且一旦发生该种异常，就必须采取某种方式进行处理。

> **除了`RuntimeException`及其子类以外，其他的`Exception`类及其子类都属于检查异常**，当程序中可能出现这类异常，**要么使用`try-catch`语句进行捕获，要么用`throws`子句抛出**，否则编译无法通过。

## 不受检查异常

**包括`RuntimeException`及其子类和`Error`**。

> `不受检查异常`为编译器不要求强制处理的异常，`检查异常`则是编译器要求必须处置的异常。

## try-catch 返回值

首先要注意，如果`finally`和`try`联合使用，`finally`将会在`try`结束之前执行。

1. 如果try有返回值，返回的是try里面变量值保存到局部变量中
2. JSR指令先去finally语句中执行，在返回局部变量中值
3. **如果try，finally中均有return，忽略try中的return，使用finally中的return**

## try-catch 关于AutoCloseAble接口

IO流中，需要使用try-catch进行资源的关闭，但是这样的代码会冗余，所以就在java7中增加了AutocloseAble接口，只要是实现这个接口的资源类，声明时放在try语句中，就不需要特地关闭资源。

```java
try(FileOutPutStream out = new FileOutPutStream(...)) {
    ...
} catch (Exception e) {
    e.printStackTrace();
} // 无需再在finally中添加关闭资源的代码
```

执行顺序是，先执行try中的语句，不论出没出现异常，先将资源关闭，然后捕获异常，最后执行finally